File: src/main/java/net/wdsj/mcserver/gui/common/config/GuiItemStackConfig.java
package net.wdsj.mcserver.gui.common.config;

import com.google.common.collect.Lists;
import lombok.Getter;
import mc233.cn.wdsjlib.global.config.ItemStackConfig;
import net.wdsj.servercore.common.placeholder.PlaceholderManager;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/**
 * @author Arthur
 * @version 1.0
 * @date 2018/8/27 21:54
 */
@Getter
public class GuiItemStackConfig {

    private String id;
    private String material;
    private String durability;
    private String amount;
    private String name;
    private List<String> lore = new ArrayList<>();

    private String head;
    private String house;
    private String skin;

    private List<String> enchant = new ArrayList<>();
    private List<String> flag = new ArrayList<>();


    //TODO SHIT
    public ItemStackConfig build(Object handler, Map<String, String> map) {
       // System.out.printf(" debug > material: %s name: %s lore: %s%n", material, name, ArrayUtils.toString(lore));
        ItemStackConfig itemStackConfig = new ItemStackConfig();
        for (Map.Entry<String, String> entry : map.entrySet()) {
            String foo = "$" + entry.getKey();
            if (id != null)
                id = id.replace(foo, entry.getValue());
            if (material != null)
                material = material.replace(foo, entry.getValue());
            if (durability != null)
                durability = durability.replace(foo, entry.getValue());
            if (amount != null)
                amount = amount.replace(foo, entry.getValue());
            if (name != null)
                name = name.replace(foo, entry.getValue());
            if (head != null)
                head = PlaceholderManager.replace(handler, head.replace(foo, entry.getValue()));
            if (house != null)
                house = house.replace(foo, entry.getValue());
            if (skin != null)
                skin = PlaceholderManager.replace(handler, skin.replace(foo, entry.getValue()));
            if (lore != null)
                itemStackConfig.setLore(Lists.transform(lore, s -> s.replace(foo, entry.getValue())));
            if (enchant != null)
                itemStackConfig.setEnchant(Lists.transform(enchant, s -> s.replace(foo, entry.getValue())));
            if (flag != null)
                itemStackConfig.setFlag(Lists.transform(flag, s -> s.replace(foo, entry.getValue())));

        }
        itemStackConfig.setId(id == null ? null : Integer.parseInt(id));
        itemStackConfig.setMaterial(material == null ? null : material);
        itemStackConfig.setDurability(durability == null ? null : Integer.parseInt(durability));
        itemStackConfig.setAmount(amount == null ? null : Integer.parseInt(amount));
        itemStackConfig.setName(name == null ? null : name);
        itemStackConfig.setHead(head == null ? null : head);
        itemStackConfig.setHouse(house == null ? null : house);
        itemStackConfig.setSkin(skin == null ? null : skin);

        itemStackConfig.setLore(lore);
        itemStackConfig.setEnchant(enchant);
        itemStackConfig.setFlag(flag);
        return itemStackConfig;
    }

}

File: src/main/java/net/wdsj/mcserver/gui/common/config/GuiItemStackConfig.kt
package net.wdsj.mcserver.gui.common.config

import mc233.cn.wdsjlib.global.config.ItemStackConfig
import net.wdsj.servercore.common.placeholder.PlaceholderManager
import net.wdsj.servercore.compatible.XMaterial
import net.wdsj.servercore.utils.extensions.isArmor
import net.wdsj.servercore.utils.extensions.isPotion
import net.wdsj.servercore.utils.extensions.isTool
import net.wdsj.servercore.utils.extensions.isWeapon
import java.util.stream.Collectors

/**
 * @author Arthur
 * @version 1.0
 * @date 2018/8/27 21:54
 */
//TODO SHIT, AUTO CONVERT TO KOTLIN
class GuiItemStackConfig {
    public var id: String? = null
    public var material: String? = null
    public var durability: String? = null
    public var amount: String? = null
    public var name: String? = null
    public var lore: ArrayList<String> = ArrayList()
    public val format = true
    public var head: String? = null
    public var house: String? = null
    public var skin: String? = null
    public var enchant: ArrayList<String> = ArrayList()
    public var flag: ArrayList<String> = ArrayList()

    fun build(handler: Any, map: Map<String, String?>?): ItemStackConfig {
        val itemStackConfig = ItemStackConfig()
        map?.let {
            for ((key, value) in it) {
                val foo = "$$key"
                if (id != null) id = id!!.replace(foo, value!!)
                if (material != null) material = material!!.replace(foo, value!!)
                if (durability != null) durability = durability!!.replace(foo, value!!)
                if (amount != null) amount = amount!!.replace(foo, value!!)
                if (name != null) name = name!!.replace(foo, value!!)
                if (head != null) head = PlaceholderManager.replace(handler, head!!.replace(foo, value!!))
                if (house != null) house = house!!.replace(foo, value!!)
                if (skin != null) {
                    skin = PlaceholderManager.replace(handler, skin!!.replace(foo, value!!))
                }

                lore = lore.stream().map { s: String -> s.replace(foo, value!!) }
                    .collect(Collectors.toList()) as ArrayList<String>

                enchant = enchant.stream().map { s: String -> s.replace(foo, value!!) }
                    .collect(Collectors.toList()) as ArrayList<String>

                flag = flag.stream().map { s: String -> s.replace(foo, value!!) }
                    .collect(Collectors.toList()) as ArrayList<String>
            }

        }
        if (format) {
            material?.let {
                val xMaterial = XMaterial.matchXMaterial(it)
                if (xMaterial.isPresent) {
                    val v = xMaterial.get()
                    flag.add("HIDE_ENCHANTMENTS")
                    if (v.isWeapon() || v.isArmor() || v.isTool()) {
                        flag.add("HIDE_ATTRIBUTES")
                    }
                    if (v.isPotion()){
                        flag.add("HIDE_POTION_EFFECTS")
                    }
                }
            }
        }
        itemStackConfig.id = if (id == null) null else id!!.toInt()
        itemStackConfig.material = if (material == null) null else material
        itemStackConfig.durability = if (durability == null) null else durability!!.toInt()
        itemStackConfig.amount = if (amount == null) null else amount!!.toInt()
        itemStackConfig.name = if (name == null) null else name
        itemStackConfig.head = if (head == null) null else head
        itemStackConfig.house = if (house == null) null else house
        itemStackConfig.skin = if (skin == null) null else skin
        itemStackConfig.lore = lore
        itemStackConfig.enchant = enchant
        itemStackConfig.flag = flag
        return itemStackConfig
    }
}